source/common/quic/envoy_quic_server_stream.cc

```diff
@@ -257,19 +257,28 @@ void EnvoyQuicServerStream::maybeDecodeTrailers() {
 }
 
 bool EnvoyQuicServerStream::OnStopSending(quic::QuicRstStreamErrorCode error) {
+  // Only called in IETF Quic to close write side.
+  ENVOY_STREAM_LOG(debug, "received STOP_SENDING with reset code={}", *this, error);
+  stats_.rx_reset_.inc();
   bool ret = quic::QuicSpdyServerStreamBase::OnStopSending(error);
-  if (read_side_closed()) {
+  if (read_side_closed() && !local_end_stream_) {
     // Treat this as a remote reset, since the stream will be closed in both directions.
     runResetCallbacks(quicRstErrorToEnvoyRemoteResetReason(error));
   }
   return ret;
 }
 
 void EnvoyQuicServerStream::OnStreamReset(const quic::QuicRstStreamFrame& frame) {
-  ENVOY_STREAM_LOG(debug, "received reset code={}", *this, frame.error_code);
+  ENVOY_STREAM_LOG(debug, "received RST_STREAM with reset code={}", *this, frame.error_code);
   stats_.rx_reset_.inc();
+  // This closes read side in both Google Quic and IETF Quic, but doesn't close write side in IETF
+  // Quic.
   quic::QuicSpdyServerStreamBase::OnStreamReset(frame);
-  runResetCallbacks(quicRstErrorToEnvoyRemoteResetReason(frame.error_code));
+  if (write_side_closed() && !local_end_stream_) {
+    // If the stream is closed in both directions and upstream hasn't ended the stream, run reset
+    // stream callback.
+    runResetCallbacks(quicRstErrorToEnvoyRemoteResetReason(frame.error_code));
```

Comment:
You're adjusting two code paths, but only 1 test was added.  Do you know how we could add test coverage for the other modified code path?

Reply:
QuicHttpIntegrationTest tests both Google QUIC and IETF QUIC, and the 2nd code path is only reachable in Google QUIC whose RST_STREAM means close both directions. That's why I added the test there instead of protocol_integration_test.

Reply:
quic_protocol_integration_test should test both IETF QUIC and Google QUIC.  Please move this test to protocol_integration_test to avoid duplication, we need this test for H2 also.

Reply:
Moved it to protocol_integration_test. Since we are not going to support gQUIC in Envoy, I won't add it's test coverage to quic_protocol_integration_test for now.